src/cryptonote_protocol/cryptonote_protocol_handler.inl

```diff
@@ -361,6 +366,312 @@ namespace cryptonote
+      return 1;
+    
+    m_core.pause_mine();
+      
+    block new_block;
+    transaction miner_tx;
+    if(parse_and_validate_block_from_blob(arg.b.block, new_block))
+    {
+      // This is a seccond notification, we must have asked for some missing tx
+      if(!context.m_requested_objects.empty())
+      {
+        // What we asked for != to what we received ..
+        if(context.m_requested_objects.size() != new_block.tx_hashes.size())
+        {
+          LOG_ERROR_CCONTEXT
+          (
+            "NOTIFY_NEW_FLUFFY_BLOCK -> request/response mismatch, " 
+            << "block = " << epee::string_tools::pod_to_hex(get_blob_hash(arg.b.block))
+            << ", requested = " << context.m_requested_objects.size() 
+            << ", received = " << new_block.tx_hashes.size()
+            << ", dropping connection"
+          );
+          
+          m_p2p->drop_connection(context);
+          m_core.resume_mine();
+          return 1;
+        }
+      }      
+      
+      std::list<blobdata> have_tx;
+      
+      // Instead of requesting missing transactions by hash like BTC, 
+      // we do it by index (thanks to a suggestion from moneromooo) because
+      // we're way cooler .. and also because they're smaller than hashes.
+      // 
+      // Also, remember to pepper some whitespace changes around to bother
+      // moneromooo ... only because I <3 him. 
+      std::vector<size_t> need_tx_indices;
+        
+      transaction tx;
+      crypto::hash tx_hash;
+
+      BOOST_FOREACH(auto& tx_blob, arg.b.txs)
+      {
+        if(parse_and_validate_tx_from_blob(tx_blob, tx))
+        {
+          try
+          {
+            if(!get_transaction_hash(tx, tx_hash))
+            {
+              LOG_PRINT_CCONTEXT_L1
+              (
+                  "NOTIFY_NEW_FLUFFY_BLOCK: get_transaction_hash failed"
+                  << ", dropping connection"
+              );
+              
+              m_p2p->drop_connection(context);
+              m_core.resume_mine();
+              return 1;
+            }
+          }
+          catch(...)
+          {
+            LOG_PRINT_CCONTEXT_L1
+            (
+                "NOTIFY_NEW_FLUFFY_BLOCK: get_transaction_hash failed"
+                << ", exception thrown"
+                << ", dropping connection"
+            );
+                        
+            m_p2p->drop_connection(context);
+            m_core.resume_mine();
+            return 1;
+          }
+          
+          // hijacking m_requested objects in connection context to patch up
+          // a possible DOS vector pointed out by @monero-moo where peers keep
+          // sending (0...n-1) transactions.
+          // If requested objects is not empty, then we must have asked for 
+          // some missing transacionts, make sure that they're all there.
+          //
+          // Can I safely re-use this field? I think so, but someone check me!
```

Comment:
I don't know. Someone would need to dive in the p2p code to see what happens when we get messages we do not expect (ie, a sync response when we did not send a sync request). I think it's OK, but I'm not sure :)